### ISSUE TITLE
Add Interface for TLS policies

### ISSUE THREAD

**Issue by windgmbh** [opnsense/core#5060]
**Important notices**

Before you add a new report, we ask you kindly to acknowledge the following:

- [x] I have read the contributing guide lines at https://github.com/opnsense/core/blob/master/CONTRIBUTING.md
- [x] I am convinced that my issue is new after having checked both open and closed issues at https://github.com/opnsense/core/issues?q=is%3Aissue

**Is your feature request related to a problem? Please describe.**
Currently, different plugins maintain separate standards regarding TLS practices. 
Different plugins allow different ways to set certain TLS versions or even specific cipher suites.

**Describe the solution you like**
In my contribution to `mail/postfix` (https://github.com/opnsense/plugins/commit/bfdc07de980df2076a8d9a2dac0044c83e65d491) I have used [Mozilla's SSL configuration guidelines](https://ssl-config.mozilla.org/) to improve the quality of TLS version and cipher specification. 
Since then, OPNsense's `mail/postfix` plugin offers a simple choice between an 'Old', 'Intermediate' and 'Modern' configuration. 
Using Postfix as an example, I have to be critical regarding my own implementation choice, because Mozilla's recommendations are hard coded into the [`main.cf` Jinja template](https://github.com/opnsense/plugins/blob/master/mail/postfix/src/opnsense/service/templates/OPNsense/Postfix/main.cf) and have to be updated on a per-plugin basis.  
It would be better if I have had a more general way of implementing good TLS practices.

Mozilla maintains their TLS recommendations [in JSON](https://github.com/mozilla/ssl-config-generator/blob/master/src/static/guidelines/5.6.json), which could trivially included as an OPNsense model that can be used in different templates.

OPNsense components and plugins could implement their TLS settings through this. 
Again with the Postfix example there would no need to implement TLS cipher suites like 

```
tls_low_cipherlist = ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-POLY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384:DHE-RSA-CHACHA20-POLY1305:ECDHE-ECDSA-AES128-SHA256:ECDHE-RSA-AES128-SHA256:ECDHE-ECDSA-AES128-SHA:ECDHE-RSA-AES128-SHA:ECDHE-ECDSA-AES256-SHA384:ECDHE-RSA-AES256-SHA384:ECDHE-ECDSA-AES256-SHA:ECDHE-RSA-AES256-SHA:DHE-RSA-AES128-SHA256:DHE-RSA-AES256-SHA256:AES128-GCM-SHA256:AES256-GCM-SHA384:AES128-SHA256:AES256-SHA256:AES128-SHA:AES256-SHA:DES-CBC3-SHA
```

instead it could work something like
```
tls_low_cipherlist = {{ OPNsense.ssl.configurations.old.openssl | join(':') }}
```

**Additional context**

In case this suggestion would be favored by the project, I may find the time to contribute this. What do you think about this?

**Comment by AdSchellevis** [comment]
Although I do like the Mozilla standards, I'm a bit struggling on how this should work in a way that it's both maintainable and doesn't break existing setups.  Ideally one would be able to choose between the Mozilla standards or "custom" in which case all settings should be mapped accordingly. 

Persisting the settings in a model wouldn't make much sense in that case, some sort of standard settings collections which one could map into the jinja template would be more logical in that case.

functionally one would be able to do something like this in the template (pseudo code):

```
{% set tls_choice = 'intermediate' %}
tls_low_cipherlist =  {{ helpers.get_default_setting('mozilla_tls', 'configurations.' + tls_choice + '. ciphers. openssl'}}|join(':') }}
```

The question is how much reuse we would eventually offer with this given the amount of work needed to implement this, building the facility, making sure it's globally used in all available components (the webgui doesn't use a new template yet for example) and keeping the development documentation aligned.

**Comment by windgmbh** [comment]
Thanks for adding your perspective to my suggestion. I think you are absolutely right, that we would need to allow configurations other than Mozilla's. (*Even though they probably cover all of the more reasonable use cases.*)

I can offer that I will write a module `core/src/opnsense/service/modules/addons/tls_policies.py` that accepts Mozilla's policies and similarly structured data. 

We could include Mozilla's policies in `opnsense/core` and allow to add community contributed TLS policies for uncovered use cases through `opnsense/plugins`.

Writing a full blown TLS policy editor for the web-based user interface is probably not worth it or necessary, then.

> The question is how much reuse we would eventually offer with this given the amount of work needed to implement this, building the facility, making sure it's globally used in all available components 

You are of course right. Mozilla offers [various `hbs`templates](https://github.com/mozilla/ssl-config-generator/tree/master/src/templates/partials) for various components that OPNsense uses, that would just need to be transformed into Jinja. 

> making sure it's globally used in all available components (the webgui doesn't use a new template yet for example) and keeping the development documentation aligned.

I think that it is better to ship most of the components with 'Mozilla Intermediate' as default, but allow reconfiguration on a per-plugin basis. The reasoning behind this is, that different components are to be expected to have different types of clients. An OPNsense operator may for example:
* Set the administrative web user interface to 'Mozilla Modern' because he does not require TLS 1.2 compatibility. 
* while using 'Mozilla Old' for Postfix, because he still wants to maintain encryption with older peers. 

I think it is not necessary that every component would use `tls_policies.py`. I would understand it as an offer towards the maintainers of OPNsense components to upstream the question of which TLS configuration is reasonable and/ or abolish the need to implement their own mechanism for user configurable TLS settings.

**Long story short:** Are you interested in a `tls_policies.py` as a basic proof of concept?
